src/osd/scrubber/scrub_queue.h

```diff
@@ -0,0 +1,199 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "include/expected.hpp"
+#include "include/utime_fmt.h"
+#include "osd/osd_types.h"
+#include "osd/osd_types_fmt.h"
+#include "osd/scrubber/scrub_queue_if.h"
+#include "osd/scrubber_common.h"
+
+#include "utime.h"
+
+
+namespace Scrub {
+class ScrubSchedListener;
+class ScrubJob;
+class SchedEntry;
+}  // namespace Scrub
+
+/**
+ * the queue of PGs waiting to be scrubbed.
+ *
+ * the following invariants hold:
+ * - there are at most two objects for each PG (one for each scrub type) in
+ *   the queue.
+ * - if a queue element is removed or white-out, the corresponding object held
+ *   by the PgScrubber will (not necessarily immediately) be marked as
+ *   'not in the queue'.
+ * - 'white-out' queue elements are never reported to the queue users.
+ */
+class ScrubQueue : public Scrub::ScrubQueueOps {
+ public:
+  ScrubQueue(CephContext* cct, Scrub::ScrubSchedListener& osds);
+  virtual ~ScrubQueue() = default;
+
+  friend class TestOSDScrub;
+  friend class ScrubQueueTestWrapper;  ///< unit-tests structure
+
+  using SchedEntry = Scrub::SchedEntry;
+  using SchedulingQueue = std::deque<SchedEntry>;
+
+  std::ostream& gen_prefix(std::ostream& out) const;
+
+  // ///////////////////////////////////////////////////
+  // the ScrubQueueOps interface:
+
+  utime_t scrub_clock_now() const override;
+
+  Scrub::sched_conf_t populate_config_params(
+      const pool_opts_t& pool_conf) const override;
+
+  void remove_entry(spg_t pgid, scrub_level_t s_or_d) final;
+
+  void cp_and_queue_target(SchedEntry t) final;
+
+  bool queue_entries(spg_t pgid, SchedEntry shallow, SchedEntry deep) final;
+
+
+  // ///////////////////////////////////////////////////
+  // outside the scope of the I/F used by the ScrubJob:
+
+  /**
+   * the main entry point for the OSD. Called in OSD::tick_without_osd_lock()
+   * to determine if there are PGs that are ready to be scrubbed, and to
+   * initiate a scrub of one of those that are ready.
+   */
+  void sched_scrub(
+      const ceph::common::ConfigProxy& config,
+      bool is_recovery_active);
+
+  /*
+   * handles a change to the configuration parameters affecting the scheduling
+   * of scrubs.
+   */
+  void on_config_times_change();
+
+ public:
+  void dump_scrubs(ceph::Formatter* f);
+
+  /**
+   * No new scrub session will start while a scrub was initiated on a PG,
+   * and that PG is trying to acquire replica resources.
+   */
+  void set_reserving_now() { a_pg_is_reserving = true; }
+  void clear_reserving_now() { a_pg_is_reserving = false; }
+  bool is_reserving_now() const { return a_pg_is_reserving; }
+
+  // resource reservation management
+
+ private:
+  bool can_inc_scrubs() const;
+
+ public:
+  bool inc_scrubs_local();
+  void dec_scrubs_local();
+  bool inc_scrubs_remote();
+  void dec_scrubs_remote();
+  void dump_scrub_reservations(ceph::Formatter* f) const;
```

Comment:
Managing scrub reservations isn't mentioned in the class comment at the top and these don't have comments.  What's the intent here?

Reply:
I have added (in the .h) a general comment regarding the roles of the "scrub-queue"  - the OSD sub-object that
handles all its scrub affairs. This specific functionality was there already, before this PR.
Note - there are multiple interfaces that were not modified in this PR (and might have their original - or
missing - full comments).